### DISCUSSION TITLE
Conditional suppression of CS8618

### DISCUSSION THREAD

**Discussion by fiseni** [dotnet/roslyn#72717] [Q&A] [answered]
Hi,

I find the CS8618 compiler warning very useful.
But, not rarely, we do define empty constructors. It's necessary for EF Core, other ORMs, mapping libraries, etc. And once you have that, then this rule becomes an annoyance. You either have to use pragma, or initialize the state to `null!` which defies the whole purpose.

Any plans on this topic? Some way to annotate constructors; or perhaps a new compiler warning that targets only private constructors?

![image](https://github.com/dotnet/roslyn/assets/24314310/ab867a8a-08f2-46ab-ba0b-56fb43b31692)

**Comment by CyrusNajmabadi** [comment, marked as answer]
> You either have to use pragma

That seems like a very good thing.  Use of this constructor would leave the object in an invalid state, and you need to add additional information to indicate that that is ok, because there is external domain knowledge being used to know it is being properly initialized.

Note: you could always create a DiagnosticSuppressor for these cases, since you have the info on whether or not this is actually safe.

**Comment by fiseni** [comment]
Thanks @CyrusNajmabadi,

Any guidance on implementing DiagnosticSuppressors? Are there any docs with examples?
Here, just as a test, I'm suppressing all CS8618 warnings. How do I activate this? What am I missing in the csproj file?
```csharp
[DiagnosticAnalyzer(LanguageNames.CSharp)]
public sealed class DiagnosticSuppressorForConstructorNullabilityWarnings : DiagnosticSuppressor
{
    public static SuppressionDescriptor SuppressionDescriptor => new SuppressionDescriptor(
        id: "SPR0001",
        suppressedDiagnosticId: "CS8618",
        justification: "This is ok!");

    public override ImmutableArray<SuppressionDescriptor> SupportedSuppressions { get; } = [SuppressionDescriptor];

    public override void ReportSuppressions(SuppressionAnalysisContext context)
    {
        foreach (var diagnostic in context.ReportedDiagnostics)
        {
            if (SuppressionDescriptor.Id.Equals(diagnostic.Id))
            {
                context.ReportSuppression(Suppression.Create(SuppressionDescriptor, diagnostic));
            }
        }
    }
}
```

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net8.0</TargetFramework>
    <ImplicitUsings>enable</ImplicitUsings>
    <Nullable>enable</Nullable>
    <EnforceExtendedAnalyzerRules>true</EnforceExtendedAnalyzerRules> // Added this
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.CodeAnalysis.Common" Version="4.9.2" />
  </ItemGroup>

</Project>
```

**Reply by CyrusNajmabadi** [reply]
You likely need to reference your code as an `<Analyzer>` reference (or maybe it's `AnalyzerReference`.  Not sure.  On phone :)

**Reply by fiseni** [reply]
Ok, fixed it by creating a NuGet package with the analyzer. Also, I think the analyzer library should target `netstandard2.0`, otherwise won't work in VS.
Thanks!

**Comment by fiseni** [comment]
For any future reader, I published a NuGet package for this. It suppresses the CS8618 warning for private constructors only.
If anyone needs it, here you go
https://github.com/fiseni/PrivateConstructorDiagnosticSuppressor

Thanks @CyrusNajmabadi for the guidance.
